<a href="https://colab.research.google.com/github/Pomalec/AI-data-structure-demo1/blob/main/1CWK100_template.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# mount your Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# change current working directory
import os
os.chdir('/content/drive/MyDrive/Colab Notebooks/')

In [3]:
# check we can see the dataset
os.path.isfile('pneumonia_raw.csv')

True

# Task 1

## Basic structure

Please use the structure below as your starting point. Once you are up and running then you are free to adjust it, but try to stick to the following basic approach: i) divide your investigation up into smaller code cells; ii) have text cells above each code cell briefly describing what is happening in the code, along with any relevant findings it produces.

Relatively quickly, you should be able to reuse some code from the labs to prepare the assignment dataset and evaluate at least one model with a basic holdout recipe. This would get you a pass. Ask for help if you're struggling, and read the later sections once you're ready to start aiming for higher marks. Good luck!

## Dataset preparation
Example text: The following code prepares the data using a simple 60/40 holdout approach with shuffling.

In [8]:
# data preparation steps

# Importing the packages we use
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Loading all the observations
observations = pd.read_csv('pneumonia_raw.csv')
print(observations.head())

# The column we are trying to predict
target_feature = 'Pneumonia'

# Separate the examples (features) from the labels (target feature)
examples = observations.drop(columns=target_feature).to_numpy()
labels = observations[target_feature].to_numpy()

# Split the data into training and testing sets
train_examples, test_examples, train_labels, test_labels = train_test_split(
    examples,
    labels,
    test_size=0.4,
    random_state=99,
    shuffle=True
)

# Initialize a Decision Tree model
model = DecisionTreeClassifier(random_state=99)

# Train the model on the training data
model.fit(train_examples, train_labels)

# Make predictions on the test data
predictions = model.predict(test_examples)

# Evaluate the model's accuracy
accuracy = accuracy_score(test_labels, predictions)
print("Accuracy:", accuracy, "(or", round(accuracy * 100, 1), "%)")

   Patient_ID  Patient_Age  Male  Xray_Brightness  Xray_Contrast  \
0  6664019195           60   1.0              0.7            0.2   
1  4433763263           18   1.0              1.3            0.7   
2  4622245411           43   1.0              1.3            0.6   
3  2777404684           38   1.0              1.0            0.3   
4  3339776374           23   1.0              1.1            0.5   

   Silhouette_Sign  Max_Consolidation_Width  Max_Consolidation_Height  \
0              174                       32                        14   
1              316                       10                        21   
2              155                       15                        20   
3              216                       21                        24   
4              191                       37                        41   

   Cavity_Presence  Fluid_Level  Air_Bronchograms Pneumonia  
0              7.8          4.2               1.1        no  
1              6.0          

## Model 1 evaluation
Example text: The following code cell evaluates a Decision Tree model. The resulting [insert name of performance metric] is ##.

**Note that completing these first two sections ("Dataset preparation" and "Model 1 evaluation") would be enough for a pass mark :)**

In [9]:
# Evaluate model 1:

# Make predictions on the test data
predictions = model.predict(test_examples)

# Evaluate the model's accuracy
accuracy = accuracy_score(test_labels, predictions)
print("Accuracy:", accuracy, "(or", round(accuracy * 100, 1), "%)")

Accuracy: 0.5641025641025641 (or 56.4 %)


## Model 1 evaluation
The following code cell evaluates a Decision Tree model. The resulting accuracy is 56.4 %.

**Note that completing these first two sections ("Dataset preparation" and "Model 1 evaluation") would be enough for a pass mark :)**

## Model 2 evaluation
Example text: The following code cell evaluates a k-Nearest Neighbours model. The resulting [insert name of performance metric] is ##.

In [ ]:
# Evaluate model 2:

# ...

## Model 3 evaluation
Example text: The following code cell evaluates a Naive Bayes model. The resulting [insert name of performance metric] is ##.

In [ ]:
# Evaluate model 3:

# ...

## Model # evaluation
**Keep repeating the basic text cell + code cell structure for every model you evaluate.**

Example text: The following code cell evaluates a ... model. The resulting [insert name of performance metric] is ##.


In [ ]:
# Evaluate model N:

# ...

## Aiming for higher marks

Broadly speaking, for higher marks you need to: evaluate extra models; evaluate the impact of relevant hyperparameters; evaluate ensembles of relevant models; take steps to improve your recipe; justify those recipe changes in the relevant text cells. See the assignment specification for full details.

As you make improvements to your code and text cells, it's no problem temporarily to leave old code/text cells in the notebook (e.g., for safety, cross-checking). But by the time you come to submit your work, try to remove your old work and leave us just with your final approach for marking.

Task 1 code that meets the requirements of the higher marking bands also meets the requirements of the lower ones. So we only need to see where your investigation ended up, and not all the intermediate steps you took to get it there.

Any questions, don't hesitate to ask.

# Task 2



> Colleague's claim: "*Instead of non-medical staff trying to extract relevant measurements from x-rays manually, we can classify the raw x-ray image data directly; it should be possible to get improved performance, and save on staff time, without any downsides*"



## Basic structure

There is pretty much complete freedom in how you structure your analysis for task 2. Please just use plenty of section headings to keep your approach clear. Switching on the table of contents (View->Table of contents) should give us a good overview of how you have structured your analysis.

Beyond that, it's a case of adding as many text and code cells as you want/need in order to complete your analysis.

Notice that there is more than one aspect to the claim. Thinking about each aspect in turn (and potential links between them) may help you to structure your work on this task.

## Section 1: ...
Analyse (some aspect of) the claim using text cells (which can include images, equations, tables, etc.)...

In [ ]:
# and perhaps using code cells / experiments as well

# ...

Make sure you discuss any code outputs/experimental findings, and make clear their relevance to the analysis.

## Section 2: ...
Analyse (some aspect of) the claim using text cells...

In [ ]:
# and perhaps using code cells / experiments as well

# ...

...

In [ ]:
# ...

...

## Section 3: ...
Analyse (some aspect of) the claim using text cells...

In [ ]:
# and perhaps using code cells / experiments as well

# ...

...

## Section #: ...
Use as many sections as you need.

You will need to cite and reference all other sources of information, ideas or materials as part of this work, and so you will have citations within the body of the notebook, and a final "References" text cell at the end. One has been left in below as a reminder. It includes a link to the MMU library guidance for using Cite Them Right Harvard.

# Conclusion

This is where you would conclude your work on both tasks.

You should be able to get some text in place for task 1 quite quickly:

Example text for task 1: This investigation used a shuffled 60/40 train/test holdout split to evaluate the following models: [insert list of models]. It found that the ... model gave the strongest performance with a [insert name of performance metric] of ##.

You could then leave this section alone for a while as you work on adding more models to your investigation and improving your recipe. But eventually you need to come back and update it to reflect your final findings.

For task 2 you need to refer back to your earlier analysis work in order to make final judgements about each aspect of the claim. On balance, is your colleague right in what they are saying, or not? And finally, overall, do you think the claim should be pursued, versus the manual measurement process you investigated in task 1?

**If you're not able to make/conclude an analysis for task 2 you can still get a good mark for a properly concluded task 1 investigation**

# References

All other sources you draw on need properly citing and referencing using Cite Them Right Harvard. This means adding citations at the relevant points in the notebook (where you use the source), as well as the corresponding references here. The MMU library guidance for Cite Them Right Harvard is available [here](https://www.mmu.ac.uk/library/referencing-and-study-support/referencing/cite-them-right-harvard).

# Submission

Once you're ready to submit, you need to compress the whole of your 1CWK100 folder as a .zip file (no other compression formats), and upload it to Moodle. There are several ways to do this, but the instructions below offer a simple step-by-step guide:

*    **Make sure the output from any code cells you have added is showing (you can select 'Runtime' -> 'Run all' if you need to)**
*    Save your notebook (select 'File' -> 'Save')
*    Find your /1CWK100/ folder via the Google Drive web interface ([https://drive.google.com/](https://https://drive.google.com/))
*    Right-click on the /1CWK100/ folder, and select 'Download'. This will cause the folder to be compressed into a .zip file and then downloaded to your local machine.
*    Upload the resulting .zip file to the 1CWK100 submission point on Moodle.
*    (If the .zip is too large for Moodle, upload it to your University OneDrive instead, and upload a link to that copy on Moodle)

**We recommend testing these steps, and asking any questions you might have, well before the final deadline.**